java/src/test/java/org/rocksdb/SequenceNumberAllocationTest.java

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+package org.rocksdb;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+@RunWith(Parameterized.class)
+public class SequenceNumberAllocationTest {
+  @Parameters(name = "{0}")
+  public static Iterable<Boolean> parameters() {
+    return Arrays.asList(
+      true,
+      false
+    );
+  }
+
+  @Parameter(0)
+  public Boolean disableWAL;
+
+  private Options options;
+  private WriteOptions writeOptions;
+  private FlushOptions flushOptions;
+
+  @ClassRule
+  public static final RocksNativeLibraryResource ROCKS_NATIVE_LIBRARY_RESOURCE = new RocksNativeLibraryResource();
+
+  @Rule
+  public TemporaryFolder dbFolder = new TemporaryFolder();
+
+  @Before
+  public void setup()
+  {
+    options = new Options().setCreateIfMissing(true);
+    writeOptions = new WriteOptions().setDisableWAL(disableWAL);
+    flushOptions = new FlushOptions();
+  }
+
+  @After
+  public void teardown()
+  {
+    options.close();
+    writeOptions.close();
+    flushOptions.close();
+  }
+
+  @Test
+  public void testSeqnoFlushAndClose() throws RocksDBException {
+    int count;
+    try (RocksDB db = RocksDB.open(options, dbFolder.getRoot().getAbsolutePath())) {
+      count = fillDatabase(db, writeOptions, 0, 15);
+      assertEquals("Expected 15 values in the database", 15, count);
+      db.flush(flushOptions);
+      db.closeE();
+    }
+
+    try (RocksDB db = RocksDB.open(options, dbFolder.getRoot().getAbsolutePath())) {
+      assertTrue("Database should not be empty", db.getLatestSequenceNumber() != 0);
+      assertEquals("Sequence number changed between closing and reopening the database", count, db.getLatestSequenceNumber());
+    }
+  }
+
+  @Test
+  public void testSeqnoCloseWithoutFlush() throws RocksDBException {
+    int count;
+    try (RocksDB db = RocksDB.open(options, dbFolder.getRoot().getAbsolutePath())) {
+        count = fillDatabase(db, writeOptions, 0, 30);
+        assertEquals("Expected 15 values in the database", 30, count);
```

Comment:
I think we are expecting 30 here but message says 15. Shall we actually introduce a new parameter expectedCount and use that parameter in the message and assertion of using 30 here?

Reply:
This should be captured by the assertion failure, I've updated the message to "Database values count mismatch".